File: src/main/java/ru/hollowhorizon/hc/client/utils/ScissorUtil.java
package ru.hollowhorizon.hc.client.utils;

import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.Minecraft;
import org.lwjgl.opengl.GL11;

import javax.annotation.Nullable;
import java.util.ArrayDeque;
import java.util.Deque;

public class ScissorUtil {
    private static final Deque<ScissorBox> STACK = new ArrayDeque<>();

    public static void start(int x, int y, int width, int height) {
        push();
        ScissorBox.fromScreenSpace(x, y, width, height).clampInside(STACK.peek()).apply();
    }

    public static void stop() {
        RenderSystem.disableScissor();
        pop();
    }

    public static void push() {
        if (GL11.glIsEnabled(GL11.GL_SCISSOR_TEST)) {
            final int[] raw = new int[4];
            GL11.glGetIntegerv(GL11.GL_SCISSOR_BOX, raw);
            STACK.push(new ScissorBox(raw[0], raw[1], raw[2], raw[3]));
            RenderSystem.disableScissor();
        }
    }

    public static void pop() {
        if (!STACK.isEmpty()) {
            STACK.pop().apply();
        }
    }

    private static final class ScissorBox {
        private static final ScissorBox INVALID = new ScissorBox(0, 0, 0, 0);
        public final int left;
        public final int bottom;
        public final int width;
        public final int height;

        private ScissorBox(int left, int bottom, int width, int height) {
            this.left = left;
            this.bottom = bottom;
            this.width = width;
            this.height = height;
        }

        public static ScissorBox fromScreenSpace(int x, int y, int width, int height) {
            var window = Minecraft.getInstance().getWindow();
            final double scale = window.getGuiScale();
            return new ScissorBox(
                    (int) (x * scale),
                    (int) ((window.getGuiScaledHeight() - y - height) * scale),
                    (int) (width * scale),
                    (int) (height * scale));
        }

        public void apply() {
            if (this != INVALID) {
                RenderSystem.enableScissor(left, bottom, width, height);
            }
        }

        private int minX() {
            return left;
        }

        private int maxX() {
            return left + width;
        }

        private int minY() {
            return bottom;
        }

        private int maxY() {
            return bottom + height;
        }

        public ScissorBox clampInside(@Nullable ScissorBox other) {
            if (other != null) {
                final int minX = Math.max(this.minX(), other.minX());
                final int maxX = Math.min(this.maxX(), other.maxX());
                final int minY = Math.max(this.minY(), other.minY());
                final int maxY = Math.min(this.maxY(), other.maxY());
                if (maxX > minX && maxY > minY) {
                    return new ScissorBox(minX, minY, maxX - minX, maxY - minY);
                }
                return INVALID;
            }
            return this;
        }
    }
}

File: src/main/java/ru/hollowhorizon/hc/client/utils/ScissorUtil.kt
/*
 * MIT License
 *
 * Copyright (c) 2024 HollowHorizon
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package ru.hollowhorizon.hc.client.utils

import com.mojang.blaze3d.systems.RenderSystem
import net.minecraft.client.Minecraft
import org.lwjgl.opengl.GL11
import java.util.*
import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.InvocationKind
import kotlin.contracts.contract
import kotlin.math.max
import kotlin.math.min

object ScissorUtil {
    private val stack = ArrayDeque<Frame>()

    fun push(x: Int, y: Int, width: Int, height: Int) =
        push(Frame(x, y, x + width, y + height))

    fun push(frame: Frame) {
        stack.addLast(frame)
        apply()
    }

    fun pop(): Frame {
        val frame = stack.removeLast()
        apply()
        return frame
    }

    @OptIn(ExperimentalContracts::class)
    inline fun suspendScissors(fn: () -> Unit) {
        contract { callsInPlace(fn, InvocationKind.EXACTLY_ONCE) }
        val frame = pop()
        fn()
        push(frame)
    }

    private fun apply() {
        val window = Minecraft.getInstance().window

        if (stack.isEmpty()) {
            RenderSystem.disableScissor()
            return
        }

        var x1 = 0
        var y1 = 0
        var x2 = window.guiScaledWidth
        var y2 = window.guiScaledHeight

        for (frame in stack) {
            x1 = max(x1, frame.x1)
            y1 = max(y1, frame.y1)
            x2 = min(x2, frame.x2)
            y2 = min(y2, frame.y2)
        }

        val scale = window.guiScale
        RenderSystem.enableScissor(
            (x1 * scale).toInt(), (window.height - scale * y2).toInt(),
            ((x2 - x1) * scale).toInt(), ((y2 - y1) * scale).toInt()
        )
    }

    class Frame(val x1: Int, val y1: Int, val x2: Int, val y2: Int)
}
